File: ios/Talise/Features/Send/SendAmountView.swift
import SwiftUI

/// Step 1: enter an amount in the user's display currency. Big centered
/// amount, secondary USDsui-equivalent line, "MAIN WALLET" pill, custom
/// numpad. No keyboard.
struct SendAmountView: View {
    @Bindable var draft: SendDraft
    var onNext: () -> Void
    var onCancel: () -> Void

    @State private var balance: BalancesDTO?

    var body: some View {
        VStack(spacing: 0) {
            header

            Spacer(minLength: 12)

            amountBlock

            Spacer(minLength: 12)

            walletPill
                .padding(.bottom, 18)

            SendNumpad(input: $draft.rawAmount)
                .padding(.horizontal, 24)
                .padding(.bottom, 12)

            reviewButton
                .padding(.horizontal, 24)
                .padding(.bottom, 18)
        }
        .background(TaliseColor.bg.ignoresSafeArea())
        .toolbar(.hidden, for: .navigationBar)
        .onAppear { Task { await loadBalance() } }
    }

    // MARK: - Header

    private var header: some View {
        HStack {
            Button(action: onCancel) {
                Image(systemName: "xmark")
                    .font(.system(size: 16, weight: .medium))
                    .foregroundStyle(TaliseColor.fgMuted)
                    .frame(width: 36, height: 36)
                    .background(Circle().fill(TaliseColor.surfaceGlass))
            }
            Spacer()
            MicroLabel(text: "Send", color: TaliseColor.fgDim).kerning(1.5)
            Spacer()
            Color.clear.frame(width: 36, height: 36)
        }
        .padding(.horizontal, 20)
        .padding(.top, 12)
    }

    // MARK: - Amount

    private var amountBlock: some View {
        VStack(spacing: 10) {
            // Symbol + number at the SAME font size so the cap-tops
            // and baselines naturally line up. Visual hierarchy comes
            // from weight (.thin symbol vs .medium number) + color
            // (fgMuted vs fg), not from a size delta — that's what
            // produced the misaligned "tiny ₦ next to giant 0" look.
            HStack(alignment: .firstTextBaseline, spacing: 6) {
                Text(draft.currency.symbol)
                    .font(TaliseFont.heading(72, weight: .thin))
                    .foregroundStyle(TaliseColor.fgMuted)
                Text(displayAmount)
                    .font(TaliseFont.heading(72, weight: .medium))
                    .kerning(-2)
                    .foregroundStyle(TaliseColor.fg)
                    .lineLimit(1)
                    .minimumScaleFactor(0.5)
                    // Smooth digit + comma transitions as the user
                    // types past thousands boundaries. SwiftUI's
                    // numericText transition crossfades the
                    // appearing/disappearing glyphs in place, so the
                    // comma slides in cleanly when "999" → "1,000".
                    .contentTransition(.numericText())
                    .animation(.snappy(duration: 0.18), value: displayAmount)
            }

            Text(usdsuiSecondary)
                .font(TaliseFont.mono(13, weight: .light))
                .foregroundStyle(TaliseColor.fgDim)
                .contentTransition(.numericText())
                .animation(.snappy(duration: 0.18), value: usdsuiSecondary)

            // Cross-border only: show what the recipient actually
            // receives in *their* currency. Hidden for same-currency
            // sends (recipientReceiveLine == nil) so the single-currency
            // layout is untouched.
            if let recipientLine = recipientReceiveLine {
                HStack(spacing: 5) {
                    Image(systemName: "arrow.turn.down.right")
                        .font(.system(size: 10, weight: .medium))
                        .foregroundStyle(TaliseColor.accent)
                    Text(recipientLine)
                        .font(TaliseFont.mono(12, weight: .light))
                        .foregroundStyle(TaliseColor.fgMuted)
                        .contentTransition(.numericText())
                }
                .padding(.top, 1)
                .animation(.snappy(duration: 0.18), value: recipientLine)
            }

            if exceedsBalance {
                MicroLabel(
                    text: "OVER AVAILABLE BALANCE",
                    color: TaliseColor.danger
                )
                .kerning(1.5)
                .padding(.top, 2)
            }
        }
        .padding(.horizontal, 24)
    }

    /// What we render inside the big number. Formats the integer part
    /// with thousand-separator commas while preserving the raw decimal
    /// the user has typed — so mid-typing states like "12." stay as
    /// "12." (with the dangling dot) and "1234.5" reads "1,234.5".
    private var displayAmount: String {
        let raw = draft.rawAmount
        if raw.isEmpty { return "0" }
        // Locate the decimal point (if any) and group only the digits
        // to its LEFT. The right side is purely user-controlled — we
        // never reformat what they've typed past the dot.
        if let dotIdx = raw.firstIndex(of: ".") {
            let intPart = String(raw[raw.startIndex..<dotIdx])
            let fracPart = String(raw[raw.index(after: dotIdx)...])
            return "\(groupDigits(intPart)).\(fracPart)"
        }
        return groupDigits(raw)
    }

    /// Insert thousands-separator commas into a pure-digit integer
    /// string. Returns the input unchanged for strings ≤ 3 chars or
    /// for any string containing non-digits (defensive — the input
    /// here is the user's typed amount, but bracket against surprises).
    private func groupDigits(_ s: String) -> String {
        guard s.count > 3, s.allSatisfy({ $0.isNumber }) else { return s }
        var out = ""
        for (i, ch) in s.reversed().enumerated() {
            if i > 0 && i % 3 == 0 { out.append(",") }
            out.append(ch)
        }
        return String(out.reversed())
    }

    /// USDsui equivalent of what's typed, formatted as "1,234.56 USDsui".
    /// Shows "0.00 USDsui" before the user enters anything so the layout
    /// doesn't shift on the first keypress.
    private var usdsuiSecondary: String {
        let amt = typedAmountUsdsui
        let fmt = NumberFormatter()
        fmt.numberStyle = .decimal
        fmt.minimumFractionDigits = 2
        fmt.maximumFractionDigits = 2
        let body = fmt.string(from: NSNumber(value: amt)) ?? "0.00"
        return "\(body) USDsui"
    }

    private var typedAmountUsdsui: Double {
        guard let typed = Double(draft.rawAmount), typed > 0 else { return 0 }
        let rate = CurrencySettings.shared.rates[draft.currency.code] ?? 1
        guard rate > 0 else { return 0 }
        return typed / rate
    }

    /// "Recipient gets ¥15,000" line — present only for cross-border
    /// sends (different recipient currency). Returns nil for same-
    /// currency sends so the secondary block collapses to the existing
    /// single line. Mirrors the post-spread receive figure shown on the
    /// locked-quote block in Review.
    private var recipientReceiveLine: String? {
        guard let recv = draft.liveRecipientReceiveLocal() else { return nil }
        let amount = TaliseCurrency.recipientSymbolic(recv.amount, currency: recv.currency)
        return "Recipient gets \(amount)"
    }

    private var exceedsBalance: Bool {
        guard let have = balance?.usdsui else { return false }
        let amt = typedAmountUsdsui
        return amt > 0 && amt > have
    }

    // MARK: - Wallet pill

    private var walletPill: some View {
        HStack(spacing: 8) {
            Circle()
                .fill(TaliseColor.accent)
                .frame(width: 6, height: 6)
            Text("MAIN WALLET")
                .font(TaliseFont.mono(10, weight: .light))
                .kerning(1.5)
                .foregroundStyle(TaliseColor.fg)
            if let avail = availableLocal {
                Text("·")
                    .font(TaliseFont.mono(10, weight: .light))
                    .foregroundStyle(TaliseColor.fgDim)
                Text(avail)
                    .font(TaliseFont.mono(10, weight: .light))
                    .foregroundStyle(TaliseColor.fgMuted)
            }
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 8)
        .background(Capsule().fill(TaliseColor.surfaceGlass))
        .overlay(
            Capsule().stroke(TaliseColor.line, lineWidth: 0.5)
        )
    }

    private var availableLocal: String? {
        guard let usdsui = balance?.usdsui else { return nil }
        return TaliseFormat.local2(usdsui)
    }

    // MARK: - Review button

    private var reviewButton: some View {
        Button(action: handleNext) {
            Text("Review")
                .font(TaliseFont.heading(16, weight: .medium))
                .foregroundStyle(TaliseColor.bg)
                .frame(maxWidth: .infinity)
                .frame(height: 56)
                .background(canAdvance ? TaliseColor.fg : TaliseColor.fg.opacity(0.35))
                .clipShape(Capsule())
        }
        .disabled(!canAdvance)
    }

    private var canAdvance: Bool {
        typedAmountUsdsui > 0 && !exceedsBalance
    }

    private func handleNext() {
        guard canAdvance else { return }
        draft.amountUsdsui = typedAmountUsdsui
        onNext()
    }

    // MARK: - Balance

    private func loadBalance() async {
        do {
            balance = try await APIClient.shared.get("/api/balances")
        } catch {
            balance = nil
        }
    }
}

File: android/app/src/main/java/io/talise/app/feature/send/SendAmountView.kt
package io.talise.app.feature.send

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.SubdirectoryArrowRight
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.produceState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.talise.app.core.model.BalancesDTO
import io.talise.app.core.net.ApiClient
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType

/**
 * Step 1: enter an amount in the user's display currency — iOS
 * `SendAmountView`. Big centered amount, secondary USDsui-equivalent line,
 * "MAIN WALLET" pill, custom numpad. No keyboard.
 */
@Composable
fun SendAmountView(
    draft: SendDraft,
    onNext: () -> Unit,
    onCancel: () -> Unit,
) {
    val balance by produceState<BalancesDTO?>(initialValue = null) {
        value = runCatching { ApiClient.api.balances() }.getOrNull()
    }

    val typedAmountUsdsui = run {
        val typed = draft.rawAmount.toDoubleOrNull() ?: 0.0
        if (typed <= 0) 0.0 else {
            val rate = SendFx.rate(draft.currency.code)
            if (rate > 0) typed / rate else 0.0
        }
    }
    val exceedsBalance = run {
        val have = balance?.usdsui
        have != null && typedAmountUsdsui > 0 && typedAmountUsdsui > have
    }
    val canAdvance = typedAmountUsdsui > 0 && !exceedsBalance

    Column(Modifier.fillMaxSize().background(TaliseColors.bg)) {
        // Header
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 20.dp).padding(top = 12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            GlassCircleButton(Icons.Filled.Close, onClick = onCancel, iconSize = 15.dp)
            Spacer(Modifier.weight(1f))
            SendMicroLabel("Send", color = TaliseColors.fgMuted, kerning = 2.0)
            Spacer(Modifier.weight(1f))
            Spacer(Modifier.size(38.dp))
        }

        Spacer(Modifier.weight(1f))

        // Amount block
        Column(
            Modifier.fillMaxWidth().padding(horizontal = 24.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            val display = displayAmountString(draft.rawAmount)
            // Symbol + number in ONE composed text so a width-driven
            // scale-down shrinks BOTH in lockstep. The symbol renders
            // smaller + muted; hierarchy comes from weight and color.
            val chars = display.length + draft.currency.symbol.length
            val scale = if (chars <= 8) 1f else (8f / chars).coerceAtLeast(0.4f)
            val amountText = buildAnnotatedString {
                withStyle(
                    SpanStyle(
                        fontSize = (56 * scale).sp,
                        fontWeight = FontWeight.Thin,
                        color = TaliseColors.fgMuted,
                    ),
                ) { append(draft.currency.symbol) }
                withStyle(
                    SpanStyle(fontSize = (56 * scale).sp, fontWeight = FontWeight.Thin),
                ) { append(" ") }
                withStyle(
                    SpanStyle(
                        fontSize = (72 * scale).sp,
                        fontWeight = FontWeight.Medium,
                        color = TaliseColors.fg,
                        letterSpacing = (-2).sp,
                    ),
                ) { append(display) }
            }
            Text(amountText, style = TaliseType.heading(72.sp, FontWeight.Medium), maxLines = 1)

            Text(
                "${sendFmt(typedAmountUsdsui, 2)} USDsui",
                style = TaliseType.mono(13.sp, FontWeight.Light),
                color = TaliseColors.fgDim,
            )

            // Cross-border only: show what the recipient actually receives
            // in THEIR currency. Hidden for same-currency sends.
            val receive = draft.liveRecipientReceiveLocal()
            if (receive != null) {
                Row(
                    Modifier.padding(top = 1.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(5.dp),
                ) {
                    Icon(
                        Icons.Filled.SubdirectoryArrowRight,
                        contentDescription = null,
                        tint = TaliseColors.accent,
                        modifier = Modifier.size(10.dp),
                    )
                    Text(
                        "Recipient gets ${SendCurrencies.recipientSymbolic(receive.first, receive.second)}",
                        style = TaliseType.mono(12.sp, FontWeight.Light),
                        color = TaliseColors.fgMuted,
                    )
                }
            }

            if (exceedsBalance) {
                SendMicroLabel(
                    "Over available balance",
                    color = TaliseColors.danger,
                    kerning = 1.5,
                    modifier = Modifier.padding(top = 2.dp),
                )
            }
        }

        Spacer(Modifier.weight(1f))

        // Wallet pill
        Row(
            Modifier
                .align(Alignment.CenterHorizontally)
                .padding(bottom = 18.dp)
                .glassCapsule()
                .padding(horizontal = 14.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Box(Modifier.size(6.dp).background(TaliseColors.greenMint, CircleShape))
            Text(
                "MAIN WALLET",
                style = TaliseType.mono(10.sp, FontWeight.Normal),
                letterSpacing = 1.5.sp,
                color = TaliseColors.fg,
            )
            val avail = balance?.usdsui?.let { sendLocal2(it) }
            if (avail != null) {
                Text("·", style = TaliseType.mono(10.sp, FontWeight.Light), color = TaliseColors.fgDim)
                Text(avail, style = TaliseType.mono(10.sp, FontWeight.Light), color = TaliseColors.fgMuted)
            }
        }

        SendNumpad(
            value = draft.rawAmount,
            onValueChange = { draft.rawAmount = it },
            modifier = Modifier.padding(horizontal = 24.dp).padding(bottom = 12.dp),
        )

        CapsuleButton(
            title = "Review",
            enabled = canAdvance,
            onClick = {
                if (canAdvance) {
                    draft.amountUsdsui = typedAmountUsdsui
                    onNext()
                }
            },
            modifier = Modifier.padding(horizontal = 24.dp).padding(bottom = 18.dp),
        )
    }
}
